pkg/wireguard/agent/agent.go

```diff
@@ -0,0 +1,352 @@
+	"github.com/cilium/cilium/pkg/sysctl"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+const (
+	listenPort      = 51871
+	IfaceName       = "cilium_wg0"
+	PrivKeyFilename = "cilium_wg0.key"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+
+type Agent struct {
+	lock.RWMutex
+	wgClient         *wgctrl.Client
+	listenPort       int
+	privKey          wgtypes.Key
+	wireguardV4CIDR  *net.IPNet
+	wireguardV6CIDR  *net.IPNet
+	pubKeyByNodeName map[string]string // nodeName => pubKey
+	restoredPubKeys  map[string]struct{}
+}
+
+func NewAgent(privKeyPath string, wgV4Net, wgV6Net *net.IPNet) (*Agent, error) {
+	key, err := loadOrGeneratePrivKey(privKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	node.SetWireguardPubKey(key.PublicKey().String())
+
+	wgClient, err := wgctrl.New()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Agent{
+		wgClient:         wgClient,
+		privKey:          key,
+		wireguardV4CIDR:  wgV4Net,
+		wireguardV6CIDR:  wgV6Net,
+		listenPort:       listenPort,
+		pubKeyByNodeName: map[string]string{},
+		restoredPubKeys:  map[string]struct{}{},
+	}, nil
+}
+
+// Close is called when the agent stops
+func (a *Agent) Close() error {
+	return a.wgClient.Close()
+}
+
+// Init is called after we have obtained a local Wireguard IP
+func (a *Agent) Init() error {
+	link := &netlink.Wireguard{LinkAttrs: netlink.LinkAttrs{Name: IfaceName}}
+	err := netlink.LinkAdd(link)
+	if err != nil && !errors.Is(err, unix.EEXIST) {
+		return err
+	}
+
+	if err := sysctl.Write(fmt.Sprintf("net.ipv4.conf.%s.rp_filter", IfaceName), "2"); err != nil {
+		return nil
+	}
+
+	type param struct {
+		ip     *net.IPNet
+		family int
+	}
+	params := []param{}
+	if option.Config.EnableIPv4 {
+		ip := &net.IPNet{
+			IP:   node.GetWireguardIPv4(),
+			Mask: a.wireguardV4CIDR.Mask,
+		}
+		params = append(params, param{ip: ip, family: netlink.FAMILY_V4})
+	}
+	if option.Config.EnableIPv6 {
+		ip := &net.IPNet{
+			IP:   node.GetWireguardIPv6(),
+			Mask: a.wireguardV6CIDR.Mask,
+		}
+		params = append(params, param{ip: ip, family: netlink.FAMILY_V6})
+	}
+	for _, p := range params {
+		// Removes stale IP addresses from wg device
+		addrs, err := netlink.AddrList(link, p.family)
+		if err != nil {
+			return err
+		}
+		for _, addr := range addrs {
+			if !cidr.NewCIDR(addr.IPNet).Equal(cidr.NewCIDR(p.ip)) {
+				if err := netlink.AddrDel(link, &addr); err != nil {
+					return fmt.Errorf("failed to remove stale wg ip: %w", err)
+				}
+			}
+		}
+
+		err = netlink.AddrAdd(link, &netlink.Addr{IPNet: p.ip})
+		if err != nil && !errors.Is(err, unix.EEXIST) {
+			return err
+		}
+	}
+
+	cfg := &wgtypes.Config{
+		PrivateKey:   &a.privKey,
+		ListenPort:   &a.listenPort,
+		ReplacePeers: false,
+	}
+	if err := a.wgClient.ConfigureDevice(IfaceName, *cfg); err != nil {
```

Comment:
No need for this to be a pointer, since it's dereferenced right away anyway, correct?